geocity/views.py

```diff
@@ -112,27 +156,91 @@ def get_context_data(self, **kwargs):
 
     def done(self, form_list, **kwargs):
         permits_services.store_tags_in_session(self.request)
-        return super(CustomLoginView, self).done(form_list, **kwargs)
 
-    def get_success_url(self):
+        # Redirect to the duo authentication page
+        if self.duo_mfa:
+            return redirect(self.prompt_uri)
+        return super(CustomTwoFactorLoginView, self).done(form_list, **kwargs)
 
+    def get_success_url(self):
         qs_dict = parse.parse_qs(self.request.META["QUERY_STRING"])
 
-        url_value = (
-            qs_dict["next"][0]
-            if "next" in qs_dict
-            else reverse("permits:permit_request_select_administrative_entity")
-        )
-
-        if "next" in qs_dict:
-            qs_dict.pop("next")
+        url_value = get_redirect_url_after_login(qs_dict)
 
         return (
             reverse("two_factor:profile")
             if settings.ENABLE_2FA and not self.request.user.totpdevice_set.exists()
             else url_value
         )
 
+    def process_step(self, form):
+        # Override only this step
+        if self.steps.current == "auth":
+            # Get username
+            username = form.is_valid() and form.user_cache
+
+            # Get user, duo_mfa and duo_client
+            user, self.duo_mfa, duo_client = create_duo_client(username, self.request)
+
+            if self.duo_mfa:
+                # Health check
+                try:
+                    duo_client.health_check()
+                except:
+                    raise Exception("Duo n'est pas disponible pour le moment.")
+
+                # Generate the actual state
+                state = duo_client.generate_state()
+
+                # Get query_string. Used to know if there is a "next" in the request
+                qs_dict = parse.parse_qs(self.request.META["QUERY_STRING"])
+
+                # Store state, username and qs_disct in session
+                self.request.session["state"] = state
+                self.request.session["username"] = user.username
+                self.request.session["qs_dict"] = qs_dict
+
+                # Create url to duo authentication page. Used in self.done()
+                self.prompt_uri = duo_client.create_auth_url(user.username, state)
+            else:
+                # When duo isn't activated, use django 2fa
```

Comment:
Then the user continues either to Django 2FA ou basic login ? Please explain in comment